Replace remaining LSB protocol implementations with ordinary functions

athos opened the pull request:
This is the last step of #289.

This PR adds the  following implementations of LSB protocols as ordinary functions:

- reader
  - `DataInput` (defined in `cljam.io.util.lsb.data-io`)
  - `InputStream` (defined in `cljam.io.util.lsb.io-stream`)
- writer
  - `OutputStream` (defined in `cljam.io.util.lsb.io-stream`)

The `cljam.io.util.lsb` namespace has some other implementations as well, but they are no longer in use.

The following table shows the implementations used before and after the change:

- reader
  | | before change | after change |
  | :--- | :--- | :--- |
  | `cljam.io.bam.core` | `DataInputStream` (`DataInput`) | `cljam.io.util.lsb.data-io` |
  | `cljam.io.bam.reader` | `DataInputStream` (`DataInput`) | `cljam.io.util.lsb.data-io` |
  | `cljam.io.bam-index.reader` | `FileInputStream` (`InputStream`) | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bcf.reader` | `BGZFInputStream` (`InputStream`) | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bigwig` | `RandomAccessFile` (`DataInput`) | `cljam.io.util.lsb.data-io` |
  | `cljam.io.csi` | `DataInputStream` (`DataInput`) | `cljam.io.util.lsb.data-io` |
  | `cljam.io.tabix` | `DataInputStream` (`DataInput`) | `cljam.io.util.lsb.data-io` |

- writer
  | | before change| after change |
  | :--- | :--- | :--- |
  | `cljam.algo.convert` | `ByteBuffer` | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bam.encoder` | `DataOutputStream` | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bam.writer` | `DataOutputStream` | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bam-index.writer` | `DataOutputStream` | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.bcf.writer` | `DataOutputStream` | `cljam.io.util.lsb.io-stream` |
  | `cljam.io.csi` | `DataOutputStream` | `cljam.io.util.lsb.io-stream` |

I've confirmed that this change doesn't have significant negative impact on the performance of the related readers/writers (This PR also includes a tweak of the benchmark code for the BAM reader to make it more stable and reliable by reducing extra overhead of sequence allocation in the benchmark code itself):

| | before change | after change |
| :--- | ----: | ---: |
| BAM reader | `5.835351 s` | `5.677128 s` |
| BAM writer | `12.149866 s` | `12.076689 s` |
| BCF reader | `6.469314 s` | `6.468553 s` |
| BCF writer | `1.641983 s` | `1.676683 s` |
| `cljam.algo.convert` | `182.333152 ms` | `180.971159 ms` |
| BigWig reader | `84.117612 μs` | `83.471144 μs` |
| CSI reader | `22.210736 μs` | `21.966407 μs` |
| CSI writer | `127.886597 μs` | `127.627769 μs` |
| Tabix reader | `24.434094 μs` | `25.902402 μs` |

See below for more about the benchmarking details:

<details><summary>benchmarking details</summary>

For those with existing benchmark code, I benchmarked them using the following `lein libra` commands:

| target | benchmarking command |
| :--- | :--- |
| BAM reader | `lein libra :only cljam.io.sam-bench/decode-bam-alignment-long-bench` |
| BAM writer | `lein libra :only cljam.io.sam-bench/encode-alignment-long-bench` |
| BCF reader | `lein libra :only cljam.io.vcf-bench/decode-variant-large-bcf-bench` |
| BCF writer | `lein libra :only cljam.io.vcf-bench/encode-variant-large-bcf-bench` |
| `cljam.algo.convert` | `lein libra :only cljam.algo.convert-bench` |

For the others, I used the following tiny benchmark code:

```clojure
(require '[criterium.core :as cr]))

;; BigWig reader
(require '[cljam.io.bigwig :as bigwig])
(cr/quick-bench
 (with-open [r (bigwig/reader "test-resources/bigwig/test-bedgraph.bigWig")]
   (dorun (bigwig/read-tracks r))))

(require '[cljam.io.csi :as csi])
;; CSI reader
(cr/quick-bench (csi/read-index "test-resources/csi/test-v4_3-complex.vcf.gz.csi"))
;; CSI writer
(let [idx (csi/read-index "test-resources/csi/test-v4_3-complex.vcf.gz.csi")]
  (cr/quick-bench (csi/write-index "out.csi" idx)))

;; Tabix reader
(require '[cljam.io.tabix :as tbi])
(cr/quick-bench (tbi/read-index "test-resources/tabix/test.gtf.gz.tbi"))
```
</details>

alumi reviewed (approved):
LGTM 👍 
Now that `cljam.io.util.lsb` is no longer used, we can either remove the ns and cut a new release `0.9.0`, or deprecate functions in it and keep the version `0.8.5`.
I've searched our internal repositories and found just one function depending on the ns, so I think it's almost safe to remove it. Could you tell me what you think about this?

athos commented:
I’d prefer more graceful deprecation, so I've added the `:deprecated` attr to `cljam.io.util.lsb`.

I also added another commit to add notes that warn use of the new lsb namespaces to discourage undesired abuse of them from outside cljam in the future.

niyarin reviewed (approved):
LGTM 👍
